crates/task-impls/src/quorum_proposal.rs

```diff
@@ -80,32 +66,184 @@ struct ProposalDependencyHandle<TYPES: NodeType> {
+        view: TYPES::Time,
+        event_stream: &Sender<Arc<HotShotEvent<TYPES>>>,
+        commit_and_metadata: CommitmentAndMetadata<TYPES::BlockPayload>,
+        proposal_cert: Option<ViewChangeEvidence<TYPES>>,
+    ) -> Result<()> {
+        self.check_leader(view)?;
+        let consensus = self.consensus.read().await;
+        let (leaf, state) = Self::find_parent_leaf_and_state(&consensus)?;
+        let reached_decided = leaf.get_view_number() == consensus.last_decided_view;
+
+        let parent_leaf = leaf.clone();
+        let original_parent_hash = parent_leaf.commit();
+        let mut next_parent_hash = original_parent_hash;
+
+        // Walk back until we find a decide
+        if !reached_decided {
+            debug!(
+                "We have not reached decide from view {:?}",
+                self.view_number
+            );
+            while let Some(next_parent_leaf) = consensus.saved_leaves.get(&next_parent_hash) {
+                if next_parent_leaf.get_view_number() <= consensus.last_decided_view {
+                    break;
+                }
+                next_parent_hash = next_parent_leaf.get_parent_commitment();
+            }
+            debug!("updated saved leaves");
+            // TODO do some sort of sanity check on the view number that it matches decided
+        }
+
+        let block_header = TYPES::BlockHeader::new(
+            &state,
+            &consensus.instance_state,
+            &parent_leaf,
+            commit_and_metadata.commitment,
+            commit_and_metadata.metadata.clone(),
+        )
+        .await;
+
+        let upgrade_cert = if self
+            .upgrade_cert
+            .as_ref()
+            .is_some_and(|cert| cert.view_number == view)
+        {
+            // If the cert view number matches, set upgrade_cert to self.upgrade_cert
+            // and set self.upgrade_cert to None.
+            //
+            // Note: the certificate is discarded, regardless of whether the vote on the proposal succeeds or not.
+            self.upgrade_cert.clone()
+        } else {
+            // Otherwise, set upgrade_cert to None.
+            None
+        };
+
+        // We only want the proposal to be attached if any of them are valid.
+        let proposal_certificate = proposal_cert
+            .as_ref()
+            .filter(|cert| cert.is_valid_for_view(&view))
+            .cloned();
+
+        let proposal = QuorumProposal {
+            block_header,
+            view_number: view,
+            justify_qc: consensus.high_qc.clone(),
+            proposal_certificate,
+            upgrade_certificate: upgrade_cert,
+        };
+
+        let mut new_leaf = Leaf::from_quorum_proposal(&proposal);
+        new_leaf.set_parent_commitment(parent_leaf.commit());
+
+        let Ok(signature) =
+            TYPES::SignatureKey::sign(&self.private_key, new_leaf.commit().as_ref())
+        else {
+            bail!("Failed to sign leaf.commit()!");
+        };
```

Comment:
I think this could be `let signature = TYPES::SignatureKey::sign(&self.private_key, new_leaf.commit().as_ref()).context("Failed to sign leaf.commit()!")?;`